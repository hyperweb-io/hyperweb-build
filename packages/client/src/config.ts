export interface HyperwebConfig {
  name: string;
  version: string;
  chainID: string;
  node: string;
}
