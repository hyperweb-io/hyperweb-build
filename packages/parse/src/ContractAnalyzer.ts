import * as parser from '@babel/parser';
import traverse from '@babel/traverse';
import * as t from '@babel/types';

export interface AnalysisResult {
  queries: string[];
  mutations: string[];
}

export class ContractAnalyzer {
  private ast: parser.ParseResult<t.File> | null = null;

  /**
   * Analyzes a TypeScript contract class to identify query and mutation methods
   * @param code The TypeScript source code to analyze
   * @returns An object containing arrays of query and mutation method names
   */
  public analyzeFromCode(code: string): AnalysisResult {
    this.parseCode(code);
    return this.analyze();
  }

  /**
   * Parses the TypeScript code into an AST
   * @param code The TypeScript source code to parse
   */
  private parseCode(code: string): void {
    this.ast = parser.parse(code, {
      sourceType: 'module',
      plugins: ['typescript', 'classProperties', 'decorators-legacy'],
    });
  }

  /**
   * Analyzes the parsed AST to identify query and mutation methods
   * @returns An object containing arrays of query and mutation method names
   */
  private analyze(): AnalysisResult {
    if (!this.ast) {
      throw new Error('No code has been parsed. Call analyzeFromCode first.');
    }

    const queries: string[] = [];
    const mutations: string[] = [];

    traverse(this.ast, {
      ClassMethod(path) {
        // Skip static methods and constructors
        if (path.node.static || path.node.kind === 'constructor') {
          return;
        }

        const methodName = path.node.key.type === 'Identifier' ? path.node.key.name : '';
        if (!methodName) return;

        let readsState = false;
        let writesState = false;

        // Check for state access
        path.traverse({
          MemberExpression(memberPath) {
            const object = memberPath.node.object;
            const property = memberPath.node.property;

            // Check if this is a this.state access
            if (
              t.isThisExpression(object) &&
              t.isIdentifier(property) &&
              property.name === 'state'
            ) {
              // Check if this is part of an assignment
              const parent = memberPath.parentPath;
              if (
                t.isAssignmentExpression(parent.node) ||
                (t.isMemberExpression(parent.node) && 
                 t.isAssignmentExpression(parent.parentPath.node))
              ) {
                writesState = true;
              } else {
                readsState = true;
              }
            }
          },
        });

        // If a method writes to state, it's a mutation
        // If it only reads from state, it's a query
        if (writesState) {
          mutations.push(methodName);
        } else if (readsState) {
          queries.push(methodName);
        }
      },
    });

    return { queries, mutations };
  }
} 