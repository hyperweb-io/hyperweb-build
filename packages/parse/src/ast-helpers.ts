import * as t from '@babel/types';

export class ASTHelpers {
  /**
   * Checks if a node represents a member expression accessing this.state at any depth
   */
  public static isStateMemberExpression(node: t.Node): boolean {
    if (!t.isMemberExpression(node)) return false;
    const { object, property } = node;
    if (t.isThisExpression(object) && t.isIdentifier(property) && property.name === 'state') {
      return true;
    }
    if (t.isMemberExpression(object)) {
      return this.isStateMemberExpression(object);
    }
    return false;
  }

  /**
   * Checks if a node contains state member usage in assignment patterns or member expressions
   */
  public static containsStateMember(node: t.Node): boolean {
    if (t.isMemberExpression(node) && this.isStateMemberExpression(node)) {
      return true;
    }
    if (t.isObjectPattern(node)) {
      for (const prop of node.properties) {
        if (t.isRestElement(prop) && this.containsStateMember(prop.argument as t.Node)) {
          return true;
        }
        if (t.isObjectProperty(prop) && this.containsStateMember(prop.value as t.Node)) {
          return true;
        }
      }
    }
    if (t.isArrayPattern(node)) {
      for (const elem of node.elements) {
        if (elem && this.containsStateMember(elem as t.Node)) {
          return true;
        }
      }
    }
    if (t.isAssignmentPattern(node)) {
      return this.containsStateMember(node.left as t.Node);
    }
    if (t.isRestElement(node)) {
      return this.containsStateMember(node.argument as t.Node);
    }
    return false;
  }

  /**
   * Analyzes the type of an object node
   */
  public static analyzeObjectType(objectNode: t.Expression): string | null {
    // Try to infer the type from the AST context
    if (t.isMemberExpression(objectNode)) {
      // Could be this.state.someProperty - analyze the property type
      return this.inferPropertyType(objectNode);
    }

    if (t.isIdentifier(objectNode)) {
      // Could be a local variable - trace its declaration
      return this.inferVariableType(objectNode);
    }

    return null;
  }

  /**
   * Infers the type of a property access
   */
  public static inferPropertyType(memberExpr: t.MemberExpression): string | null {
    // This would require type inference from the AST
    // For now, return null to be conservative
    return null;
  }

  /**
   * Infers the type of a variable
   */
  public static inferVariableType(identifier: t.Identifier): string | null {
    // This would require tracking variable declarations and their types
    // For now, return null to be conservative
    return null;
  }
}
