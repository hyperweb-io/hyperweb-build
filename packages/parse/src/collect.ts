/**
 * File map type
 */
export type FileMap = Record<string, string>;

/**
 * Configuration options for source file collection
 */
export interface CollectOptions {
  /**
   * File extensions to try when resolving imports
   * @default ['.ts', '.tsx', '.js', '.jsx']
   */
  extensions?: string[];

  /**
   * Whether to include content in the result
   * @default true
   */
  includeContent?: boolean;
}

/**
 * Extract import paths from TypeScript/JavaScript content
 */
function extractImportPaths(content: string): string[] {
  const imports: string[] = [];

  // Regex patterns for different import styles
  const importPatterns = [
    /import\s+.*?from\s+['"](.*?)['"]/g, // import ... from '...'
    /import\s*\(\s*['"](.*?)['"]\s*\)/g, // import('...')
    /require\s*\(\s*['"](.*?)['"]\s*\)/g, // require('...')
    /export\s+.*?from\s+['"](.*?)['"]/g, // export ... from '...'
  ];

  for (const pattern of importPatterns) {
    let match;
    while ((match = pattern.exec(content)) !== null) {
      const importPath = match[1];
      if (importPath && isLocalImport(importPath)) {
        imports.push(importPath);
      }
    }
  }

  return imports;
}

/**
 * Check if an import path is local (not from node_modules)
 */
function isLocalImport(importPath: string): boolean {
  return (
    importPath.startsWith('./') || importPath.startsWith('../') || importPath.startsWith('/')
  );
}

/**
 * Resolve import path relative to current directory
 */
function resolveImportPath(importPath: string, currentDir: string): string {
  // Handle absolute paths
  if (importPath.startsWith('/')) {
    return importPath.substring(1); // Remove leading slash
  }

  // Handle relative paths
  if (importPath.startsWith('./')) {
    const cleanPath = importPath.substring(2);
    return currentDir ? `${currentDir}/${cleanPath}` : cleanPath;
  }

  if (importPath.startsWith('../')) {
    const pathParts = currentDir.split('/').filter((part) => part !== '');
    const importParts = importPath.split('/');

    // Process ../ parts
    let i = 0;
    while (i < importParts.length && importParts[i] === '..') {
      pathParts.pop(); // Go up one directory
      i++;
    }

    // Add remaining import parts
    const remainingParts = importParts.slice(i);
    return [...pathParts, ...remainingParts].join('/');
  }

  // Handle direct paths (shouldn't happen for local imports but just in case)
  return currentDir ? `${currentDir}/${importPath}` : importPath;
}

/**
 * Get parent directory path
 */
function getParentPath(path: string): string {
  return path.split('/').slice(0, -1).join('/');
}

/**
 * Collect source files and their dependencies
 *
 * @param entryFile - The entry point file path
 * @param files - Array of file objects with path and content
 * @param options - Collection options
 * @returns Map of file paths to their content
 */
export function collectSourceFiles(
  entryFile: string,
  files: Array<{ path: string; content: string }>,
  options: CollectOptions = {}
): FileMap {
  const { extensions = ['.ts', '.tsx', '.js', '.jsx'], includeContent = true } = options;

  // Create file map for quick lookup
  const fileMap = Object.fromEntries(files.map((file) => [file.path, file.content]));

  const result: FileMap = {};
  const visited = new Set<string>();

  function readFile(filePath: string): string | null {
    return fileMap[filePath] ?? null;
  }

  function tryReadFile(filePath: string): { content: string; actualPath: string } | null {
    // Try exact path first
    let content = readFile(filePath);
    if (content !== null) {
      return { content, actualPath: filePath };
    }

    // Try with extensions
    for (const ext of extensions) {
      const tryPath = filePath + ext;
      const tryContent = readFile(tryPath);
      if (tryContent !== null) {
        return { content: tryContent, actualPath: tryPath };
      }
    }

    return null;
  }

  function traverse(filePath: string): void {
    if (visited.has(filePath)) return;
    visited.add(filePath);

    const fileResult = tryReadFile(filePath);
    if (!fileResult) return;

    // Store content (or empty string if not including content)
    result[fileResult.actualPath] = includeContent ? fileResult.content : '';

    // Process imports
    const imports = extractImportPaths(fileResult.content);
    for (const importPath of imports) {
      if (isLocalImport(importPath)) {
        const resolvedPath = resolveImportPath(
          importPath,
          getParentPath(fileResult.actualPath)
        );
        traverse(resolvedPath);
      }
    }
  }

  traverse(entryFile);
  return result;
}
