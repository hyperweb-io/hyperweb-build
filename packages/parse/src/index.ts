export * from './ContractAnalyzer';
export * from './types';
