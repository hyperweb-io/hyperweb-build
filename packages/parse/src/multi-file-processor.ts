import * as parser from '@babel/parser';
import traverse from '@babel/traverse';
import * as t from '@babel/types';
import generate from '@babel/generator';

export class MultiFileProcessor {
  /**
   * Combines multiple source files into a single TypeScript file
   * @param sourceFiles Record of file paths to their source code content
   * @param entryFile The path to the entry file (optional - will auto-detect if not provided)
   * @returns Combined TypeScript code as a string
   */
  public combineSourceFiles(sourceFiles: Record<string, string>, entryFile?: string): string {
    const resolvedEntryFile = entryFile || this.detectEntryFile(sourceFiles);

    if (!sourceFiles[resolvedEntryFile]) {
      throw new Error(`Entry file ${resolvedEntryFile} not found in source files`);
    }

    // Parse all files to extract their content
    const asts = this.parseAllFiles(sourceFiles);
    const combinedParts: string[] = [];

    // Process files in dependency order, starting with non-entry files
    const entryAst = asts[resolvedEntryFile];

    // First, add all non-entry files (interfaces, types, helper classes)
    for (const [filePath, ast] of Object.entries(asts)) {
      if (filePath === resolvedEntryFile) continue;

      const content = this.extractFileContent(ast, sourceFiles[filePath], filePath, true);
      if (content.trim()) {
        combinedParts.push(content);
      }
    }

    // Finally, add the entry file (the main contract class)
    const entryContent = this.extractFileContent(
      entryAst,
      sourceFiles[resolvedEntryFile],
      resolvedEntryFile,
      false
    );
    combinedParts.push(entryContent);

    return combinedParts.join('\n\n');
  }

  /**
   * Parses all source files into ASTs
   */
  private parseAllFiles(
    sourceFiles: Record<string, string>
  ): Record<string, parser.ParseResult<t.File>> {
    const asts: Record<string, parser.ParseResult<t.File>> = {};

    for (const [filePath, content] of Object.entries(sourceFiles)) {
      try {
        asts[filePath] = parser.parse(content, {
          sourceType: 'module',
          plugins: ['typescript', 'classProperties', 'decorators-legacy'],
        });
      } catch (error: any) {
        throw new Error(`Failed to parse ${filePath}: ${error.message}`);
      }
    }

    return asts;
  }

  /**
   * Extracts relevant content from a file's AST, removing imports/exports as needed
   * @param ast The parsed AST of the file
   * @param originalCode The original source code
   * @param filePath The file path (for context)
   * @param removeExports Whether to remove export statements (true for non-entry files)
   * @returns Processed file content
   */
  private extractFileContent(
    ast: parser.ParseResult<t.File>,
    originalCode: string,
    filePath: string,
    removeExports: boolean
  ): string {
    const statements: string[] = [];

    traverse(ast, {
      // Skip import declarations
      ImportDeclaration(path) {
        path.skip();
      },

      // Handle export declarations
      ExportNamedDeclaration(path) {
        if (removeExports) {
          // For non-entry files, convert exports to regular declarations
          if (path.node.declaration) {
            statements.push(generate(path.node.declaration).code);
          }
        } else {
          // For entry file, keep exports as-is
          statements.push(generate(path.node).code);
        }
        path.skip();
      },

      ExportDefaultDeclaration(path) {
        if (removeExports) {
          // For non-entry files, convert default export to regular declaration
          statements.push(generate(path.node.declaration).code);
        } else {
          // For entry file, keep default export
          statements.push(generate(path.node).code);
        }
        path.skip();
      },

      // Keep all other top-level statements
      Statement(path) {
        // Only process top-level statements
        if (path.parentPath?.isProgram()) {
          statements.push(generate(path.node).code);
          path.skip();
        }
      },

      // Keep interface and type declarations
      TSInterfaceDeclaration(path) {
        if (path.parentPath?.isProgram()) {
          statements.push(generate(path.node).code);
          path.skip();
        }
      },

      TSTypeAliasDeclaration(path) {
        if (path.parentPath?.isProgram()) {
          statements.push(generate(path.node).code);
          path.skip();
        }
      },
    });

    return statements.join('\n');
  }

  /**
   * Automatically detects the entry file by finding files with default exported classes
   * @param sourceFiles Record of file paths to their source code content
   * @returns The path to the detected entry file
   */
  private detectEntryFile(sourceFiles: Record<string, string>): string {
    const candidateFiles: string[] = [];

    // Parse all files to find those with default exported classes
    for (const [filePath, content] of Object.entries(sourceFiles)) {
      try {
        const ast = parser.parse(content, {
          sourceType: 'module',
          plugins: ['typescript', 'classProperties', 'decorators-legacy'],
        });

        let hasDefaultExportedClass = false;

        traverse(ast, {
          // Handle export default class
          ExportDefaultDeclaration(path) {
            const declaration = path.node.declaration;
            if (t.isClassDeclaration(declaration)) {
              hasDefaultExportedClass = true;
            }
          },
          // Handle export { Class as default }
          ExportNamedDeclaration(path) {
            const specifiers = path.node.specifiers;
            for (const specifier of specifiers) {
              if (
                t.isExportSpecifier(specifier) &&
                specifier.exported.type === 'Identifier' &&
                specifier.exported.name === 'default'
              ) {
                // Find the variable declaration for the exported class
                const binding = path.scope.getBinding(specifier.local.name);
                if (binding) {
                  const declaration = binding.path.node;
                  if (
                    t.isVariableDeclarator(declaration) &&
                    t.isClassExpression(declaration.init)
                  ) {
                    hasDefaultExportedClass = true;
                  }
                }
              }
            }
          },
        });

        if (hasDefaultExportedClass) {
          candidateFiles.push(filePath);
        }
      } catch (error) {
        // Skip files that can't be parsed
        continue;
      }
    }

    if (candidateFiles.length === 0) {
      throw new Error('No files with default exported classes found in source files');
    }

    if (candidateFiles.length === 1) {
      return candidateFiles[0];
    }

    // If multiple candidates, use prioritization logic
    return this.prioritizeEntryFile(candidateFiles);
  }

  /**
   * Prioritizes entry file candidates when multiple files have default exported classes
   * @param candidates Array of file paths that contain default exported classes
   * @returns The prioritized entry file path
   */
  private prioritizeEntryFile(candidates: string[]): string {
    // Priority order for file naming patterns
    const patterns = [
      /contract\.ts$/i,
      /index\.ts$/i,
      /main\.ts$/i,
      /app\.ts$/i,
      /\.contract\.ts$/i,
    ];

    // Try each pattern in order
    for (const pattern of patterns) {
      const match = candidates.find((file) => pattern.test(file));
      if (match) {
        return match;
      }
    }

    // If no pattern matches, prefer files with "contract" in the name
    const contractFile = candidates.find((file) => file.toLowerCase().includes('contract'));
    if (contractFile) {
      return contractFile;
    }

    // If still no preference, choose the shortest path (likely to be more central)
    candidates.sort((a, b) => a.length - b.length);
    return candidates[0];
  }
}
