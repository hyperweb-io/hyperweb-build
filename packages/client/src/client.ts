import deepmerge from 'deepmerge';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import * as yaml from 'js-yaml';
import { mkdirp } from 'mkdirp';
import { dirname, resolve } from 'path';
import * as shell from 'shelljs';

import { HyperwebConfig } from './config';
import { readAndParsePackageJson } from './package';

export interface HyperwebContext {
  name?: string;
  rpc?: string;
  config?: string;
  verbose?: boolean;
  timeout?: string;
}

export const defaultHyperwebContext: Partial<HyperwebContext> = {
  name: '',
  rpc: 'http://localhost:26657',
  verbose: false,
};

export interface HyperwebClientI {
  ctx: HyperwebContext;
  version: string;
  config: HyperwebConfig;
}

export class HyperwebClient implements HyperwebClientI {
  ctx: HyperwebContext;
  version: string;
  config: HyperwebConfig;

  constructor(ctx: HyperwebContext) {
    this.ctx = deepmerge(defaultHyperwebContext, ctx);
    // TODO add semver check against net
    this.version = readAndParsePackageJson().version;
  }

  private log(str: string): void {
    // add log level
    console.log(str);
  }

  private exit(code: number): void {
    shell.exit(code);
  }

  private loadYaml(filename: string): any {
    const path = filename.startsWith('/')
      ? filename
      : resolve((process.cwd(), filename));
    const fileContents = readFileSync(path, 'utf8');
    return yaml.load(fileContents);
  }

  private saveYaml(filename: string, obj: any): any {
    const path = filename.startsWith('/')
      ? filename
      : resolve((process.cwd(), filename));
    const yamlContent = yaml.dump(obj);
    mkdirp.sync(dirname(path));
    writeFileSync(path, yamlContent, 'utf8');
  }

  public loadConfig(): void {
    this.ensureFileExists(this.ctx.config);
    this.config = this.loadYaml(this.ctx.config) as HyperwebConfig;
  }

  public saveConfig(): void {
    this.saveYaml(this.ctx.config, this.config);
  }

  public setConfig(config: HyperwebConfig): void {
    this.config = config;
  }

  public setContext(ctx: HyperwebContext): void {
    this.ctx = ctx;
  }

  private ensureFileExists(filename: string): void {
    const path = filename.startsWith('/')
      ? filename
      : resolve((process.cwd(), filename));
    if (!existsSync(path)) {
      throw new Error(`Configuration file not found: ${filename}`);
    }
  }
}