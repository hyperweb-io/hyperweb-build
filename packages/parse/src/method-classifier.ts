import * as t from '@babel/types';
import { NodePath } from '@babel/traverse';
import { ASTHelpers } from './ast-helpers';

export class MethodClassifier {
  // Track methods currently being analyzed to prevent infinite recursion
  private analyzerStack: Set<string> = new Set();

  /**
   * Determines if a method call is read-only by analyzing the method implementation
   * Uses pure AST analysis with recursion protection
   */
  public isReadOnlyMethod(methodName: string, callPath?: NodePath<t.CallExpression>): boolean {
    // Create a unique key for this analysis to prevent infinite recursion
    const analysisKey = `${methodName}:${callPath?.node.loc?.start.line || 'unknown'}`;

    if (this.analyzerStack.has(analysisKey)) {
      // We're already analyzing this method - default to mutation (conservative)
      return false;
    }

    this.analyzerStack.add(analysisKey);

    try {
      // Try to analyze the actual method implementation using AST
      if (callPath) {
        const methodAnalysis = this.analyzeMethodImplementation(methodName, callPath);
        if (methodAnalysis !== null) {
          return methodAnalysis;
        }
      }

      // If we can't find or analyze the method implementation,
      // default to mutation (conservative approach - no pattern matching)
      return false;
    } finally {
      this.analyzerStack.delete(analysisKey);
    }
  }

  /**
   * Analyzes the actual implementation of a method to determine if it's read-only
   * Returns null if the method can't be found or analyzed
   */
  private analyzeMethodImplementation(
    methodName: string,
    callPath: NodePath<t.CallExpression>
  ): boolean | null {
    // First try to find the method in the current class or accessible scope
    const methodNode = this.findMethodDefinition(methodName, callPath);
    if (methodNode) {
      return this.isMethodBodyReadOnly(methodNode);
    }

    // Try to find the method in imported modules
    const importedMethodNode = this.findImportedMethodDefinition(methodName, callPath);
    if (importedMethodNode) {
      return this.isMethodBodyReadOnly(importedMethodNode);
    }

    // Try to analyze based on the object type the method is called on
    const objectTypeAnalysis = this.analyzeMethodByObjectType(methodName, callPath);
    if (objectTypeAnalysis !== null) {
      return objectTypeAnalysis;
    }

    // If we can't analyze the method, return null to indicate unknown
    return null;
  }

  /**
   * Finds the method definition for the given method name
   */
  private findMethodDefinition(
    methodName: string,
    callPath: NodePath<t.CallExpression>
  ): NodePath<t.ClassMethod | t.ClassProperty> | null {
    // Start from the call path and traverse up to find the containing class
    let currentPath: NodePath | null = callPath;

    while (currentPath) {
      if (currentPath.isClassDeclaration() || currentPath.isClassExpression()) {
        // Found the class, now look for the method
        const classPath = currentPath as NodePath<t.ClassDeclaration | t.ClassExpression>;
        let foundMethod: NodePath<t.ClassMethod | t.ClassProperty> | null = null;

        classPath.traverse({
          ClassMethod(methodPath: NodePath<t.ClassMethod>) {
            if (
              t.isIdentifier(methodPath.node.key) &&
              methodPath.node.key.name === methodName &&
              !methodPath.node.static
            ) {
              foundMethod = methodPath;
              methodPath.stop();
            }
          },
          ClassProperty(propPath: NodePath<t.ClassProperty>) {
            if (
              t.isIdentifier(propPath.node.key) &&
              propPath.node.key.name === methodName &&
              !propPath.node.static &&
              t.isArrowFunctionExpression(propPath.node.value)
            ) {
              foundMethod = propPath;
              propPath.stop();
            }
          },
        });

        return foundMethod;
      }
      currentPath = currentPath.parentPath;
    }

    return null;
  }

  /**
   * Attempts to find method definitions in imported modules
   */
  private findImportedMethodDefinition(
    methodName: string,
    callPath: NodePath<t.CallExpression>
  ): NodePath<t.ClassMethod | t.ClassProperty> | null {
    const callee = callPath.node.callee;
    if (!t.isMemberExpression(callee)) {
      return null;
    }

    // Try to trace the object type through imports
    const objectNode = callee.object;
    if (t.isMemberExpression(objectNode)) {
      // Handle cases like this.state.someImportedObject.method()
      return this.traceImportedMethod(objectNode, methodName);
    }

    return null;
  }

  /**
   * Traces method calls through imported objects
   */
  private traceImportedMethod(
    objectNode: t.MemberExpression,
    methodName: string
  ): NodePath<t.ClassMethod | t.ClassProperty> | null {
    // This would require more sophisticated type tracking
    // For now, return null to indicate we can't analyze
    return null;
  }

  /**
   * Analyzes method calls based on the type of object they're called on
   */
  private analyzeMethodByObjectType(
    methodName: string,
    callPath: NodePath<t.CallExpression>
  ): boolean | null {
    const callee = callPath.node.callee;
    if (!t.isMemberExpression(callee)) {
      return null;
    }

    // For built-in JavaScript types, we need to analyze the context
    // instead of hardcoding patterns
    const objectAnalysis = ASTHelpers.analyzeObjectType(callee.object);
    if (objectAnalysis) {
      return this.analyzeMethodOnType(methodName, objectAnalysis, callPath);
    }

    return null;
  }

  /**
   * Analyzes whether a method on a specific type is read-only
   */
  private analyzeMethodOnType(
    methodName: string,
    objectType: string,
    callPath: NodePath<t.CallExpression>
  ): boolean | null {
    // Without hardcoded patterns, we can only analyze if we have
    // actual method implementations or type definitions
    // For now, return null to indicate we can't determine
    return null;
  }

  /**
   * Analyzes a method body to determine if it only reads state (true) or mutates it (false)
   */
  private isMethodBodyReadOnly(
    methodPath: NodePath<t.ClassMethod | t.ClassProperty>
  ): boolean {
    let hasStateWrite = false;
    let hasStateRead = false;
    const self = this; // Capture the MethodClassifier instance

    const analyzeBody = (bodyPath: NodePath) => {
      bodyPath.traverse({
        // Direct state assignment: this.state.x = value
        MemberExpression(memberPath: NodePath<t.MemberExpression>) {
          const obj = memberPath.node.object;
          const prop = memberPath.node.property;

          if (t.isThisExpression(obj) && t.isIdentifier(prop) && prop.name === 'state') {
            const parent = memberPath.parentPath;
            if (
              t.isAssignmentExpression(parent.node) ||
              (t.isMemberExpression(parent.node) &&
                t.isAssignmentExpression(parent.parentPath.node))
            ) {
              hasStateWrite = true;
            } else {
              hasStateRead = true;
            }
          }
        },

        // Assignment expressions that might involve state
        AssignmentExpression(assignPath: NodePath<t.AssignmentExpression>) {
          if (ASTHelpers.containsStateMember(assignPath.node.left as t.Node)) {
            hasStateWrite = true;
          }
        },

        // Update expressions (++, --, etc.) on state
        UpdateExpression(updatePath: NodePath<t.UpdateExpression>) {
          if (ASTHelpers.isStateMemberExpression(updatePath.node.argument as t.Node)) {
            hasStateWrite = true;
          }
        },

        // Method calls on state objects
        CallExpression(callExprPath: NodePath<t.CallExpression>) {
          const callee = callExprPath.node.callee;
          if (t.isMemberExpression(callee)) {
            if (ASTHelpers.isStateMemberExpression(callee.object)) {
              if (t.isIdentifier(callee.property)) {
                // Method calls on state objects - for pure AST analysis,
                // we analyze the context rather than the method name
                const parent = callExprPath.parentPath;

                // If the method call result is being returned or used in an expression,
                // it's likely a read operation
                if (
                  t.isReturnStatement(parent.node) ||
                  t.isVariableDeclarator(parent.node) ||
                  t.isAssignmentExpression(parent.node)
                ) {
                  hasStateRead = true;
                } else {
                  // If the method call is a standalone statement,
                  // it's more likely a mutation operation
                  hasStateWrite = true;
                }
              }
            }

            // Object.assign on state
            if (
              t.isIdentifier(callee.object) &&
              callee.object.name === 'Object' &&
              t.isIdentifier(callee.property) &&
              callee.property.name === 'assign'
            ) {
              const target = callExprPath.node.arguments[0];
              if (target && ASTHelpers.isStateMemberExpression(target as t.Node)) {
                hasStateWrite = true;
              }
            }
          }
        },
      });
    };

    // Analyze method body based on type
    if (methodPath.isClassMethod()) {
      const methodNode = methodPath.node as t.ClassMethod;
      if (methodNode.body) {
        analyzeBody(methodPath.get('body') as NodePath);
      }
    } else if (methodPath.isClassProperty()) {
      const propNode = methodPath.node as t.ClassProperty;
      if (t.isArrowFunctionExpression(propNode.value)) {
        const funcPath = methodPath.get('value') as NodePath<t.ArrowFunctionExpression>;
        if (t.isBlockStatement(propNode.value.body)) {
          analyzeBody(funcPath.get('body') as NodePath);
        } else {
          // Expression body - analyze the expression
          analyzeBody(funcPath.get('body') as NodePath);
        }
      }
    }

    // If there are any writes to state, it's not read-only
    return !hasStateWrite;
  }
}
