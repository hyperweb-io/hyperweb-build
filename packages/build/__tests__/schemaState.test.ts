import fs from 'fs/promises';
import { join, resolve } from 'path';

import { HyperwebBuild, HyperwebBuildOptions, schemaExtractorPlugin } from '../src';

const fixtureDir = resolve(join(__dirname, '/../../../__fixtures__/', 'schema-data', 'state-export'));
const outputDir = resolve(join(__dirname, '/../../../__output__/', 'schema-data', 'state-export'));

describe('HyperwebBuild', () => {
  it('builds the fixture project successfully', async () => {
    const outfile = join(outputDir, 'bundle.js');
    const schemaOutputPath = `${outfile}.schema.json`;

    const options: Partial<HyperwebBuildOptions> = {
      entryPoints: [join(fixtureDir, 'index.ts')],
      outfile,
      customPlugins: [
        schemaExtractorPlugin({
          outputPath: schemaOutputPath,
          baseDir: fixtureDir,
          include: [/\.ts$/],  // Only process TypeScript files
          exclude: [/node_modules/, /\.test\.ts$/],  // Skip node_modules and test files
        }),
      ],
    };

    await HyperwebBuild.build(options);

    // Check if the output file exists
    const outfileExists = await fs.access(outfile)
      .then(() => true)
      .catch(() => false);

    expect(outfileExists).toBe(true);

    // Check if schema file exists
    const schemafileExists = await fs.access(schemaOutputPath)
      .then(() => true)
      .catch(() => false);

    expect(schemafileExists).toBe(true);

    // Optionally, read the schema file and verify the contents
    const schemaContent = await fs.readFile(schemaOutputPath, 'utf-8');
    const schemaData = JSON.parse(schemaContent);

    // Perform checks on schema structure (example: check if 'state' exists and has properties)
    expect(schemaData).toHaveProperty('state');
    expect(schemaData.state).toHaveProperty('type', 'object');
    expect(schemaData.state).toHaveProperty('properties');

    expect(schemaData).toMatchSnapshot();
  });
});
