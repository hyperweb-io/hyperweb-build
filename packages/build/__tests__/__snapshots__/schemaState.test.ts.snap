// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HyperwebBuild Schema State builds the fixture project successfully 1`] = `
{
  "methods": [],
  "state": {
    "properties": {
      "count": {
        "type": "number",
      },
      "startCoin": {
        "properties": {
          "amount": {
            "type": "string",
          },
          "denom": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "tokens": {
        "items": {
          "properties": {
            "amount": {
              "type": "string",
            },
            "denom": {
              "type": "string",
            },
          },
          "type": "object",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
}
`;
